First version of the PEP

pablogsal opened the pull request:
<!--

Please include the PEP number in the pull request title, example:

PEP NNN: Summary of the changes made

In addition, please sign the CLA.

For more information, please read our Contributing Guidelines (CONTRIBUTING.rst)

-->

pablogsal committed:
First version of the PEP

pablogsal committed:
First version of the PEP

pablogsal commented:
CC @isidentical @lysnikolaou

isidentical commented on pep-3157.txt:
```diff
@@ -0,0 +1,247 @@
+Abstract
+========
+
+This document proposes to lift some of the restrictions originally formulated in
+:pep:`498` and to provide a formalized grammar for f-strings that can be
+integrated into parser directly. The proposed syntactic formalization of
```
Should we rephrase it by initially emphasizing that this is about formalization (instead of lifting/relaxing the f-string grammar) and then mention that through this we can lift restrictions imposed originally because of the implementation (potentially with a link pointing towards the section the changes are covered)?

pablogsal replied:
I agree, but I don't want that it seems that we are "hiding" the new stuff. For the end users the most impactful part of this PEP is, actually, lifting the restrictions so in some way that's going to be the one we should clarify more, I think.

isidentical commented on pep-3157.txt:
```diff
@@ -0,0 +1,247 @@
+
+When f-strings were originally introduced in :pep:`398` the specification was
+provided without providing a formal Grammar for f-strings. Additionally, theGG
+specification contains several restrictions that are imposed so the parsing of
```
```suggestion
When f-strings were originally introduced in :pep:`498` the specification was
provided without providing a formal grammar for f-strings. Additionally, the
specification contains several restrictions that are imposed so the parsing of
```

isidentical commented on pep-3157.txt:
```diff
@@ -0,0 +1,247 @@
+    >>> f'Magic wand { bag[\'wand\'] } string'
+    SyntaxError: f-string expression portion cannot include a backslash
+
+#. (TODO: DO WE WANT THIS?) Comments are forbidden even in multi-line f-strings::
```
Ideally a similar example would be this:
 ```py
 # Valid
func(
    a = 1, # comment
)

# Invalid
func(
    a = 1, # comment)

# Valid
func(
    a = 1, # comment)
) # <======= the parentheses above is not included 
```

But I am not super sure about what happens when either the enclosing parentheses is found inside the comment or the enclosing string literals.

```py
# Valid
f'''A complex trick: {
    bag['bag']  # recursive bags!
}'''

# Completely invalid
f'''A complex trick: {
    bag['bag']  # recursive bags!}
'''

# Completely invalid
f'''A complex trick: {
    bag['bag']  # recursive bags!}'''


# Potentially valid
f'''A complex trick: {
    bag['bag']  # recursive bags!}
}''' # <=== paren is still here, because the one above is ignored

# Potentially valid
f'''A complex trick: {
    bag['bag']  # recursive bags!}'''
}''' # <=== the triple quotes above does not close the string, this one does
```

Should we decide on this?

pablogsal replied:
> when either the enclosing parentheses is found inside the comment or the enclosing string literals.

I think the more predictable way is that everything after the comment is just ignored, which means that if there is a "}" after the comment then is just ignored and part of the comment, which means that you still need an extra "}" as in the last example. Do you see any problems with this?

lysnikolaou replied:
agree with @pablogsal, ignoring everything after the `#` seems reasonable.

lysnikolaou replied:
agree with @pablogsal, ignoring everything after the `#` seems reasonable.

isidentical replied:
Makes sense. My confusion was around `#` is completely ignored inside triple quoted strings, but considering you are starting it from an expression context and then trying to move to a string context it makes sense to `# ...` would omit everything in that line from the parser. 👍🏻

pablogsal commented:
Btw, we may want to change the grammar to the one they seem to use in LibCST if we find it more readable/better:

https://github.com/Instagram/LibCST/blob/main/native/libcst/src/parser/grammar.rs#L1401

@isidentical @lysnikolaou What do you think?

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,247 @@
+    ... }'''
+    SyntaxError: f-string expression part cannot include '#'
+
+#. (TODO: DO WE WANT THIS?) Expression portions need to wrap ``':'`` and ``'!'`` in braces::
+
+    >>> f'Useless use of lambdas: { lambda x: x*2 }'
+    SyntaxError: unexpected EOF while parsing
```
I'm somewhat torn on this, I feel that if we present the lambda as the example, support will be for not allowing it, since defining (and calling) a lambda in an f-string does not present a use case that makes sense. Are there any other examples we could use maybe?

pablogsal replied:
I added it here to discuss it together but I think this is cursed. One of the reasons is that the : defined the format specifier and we need to parse that as a string but if we allow that chunk to be tokenised normally it will never match sone of the allowed format specifiers because those are invalid combinations of alphanumeric characters.

This means that the tokenizer will always emit a FSTRING_MIDDLE and the lambda will never parse.

pablogsal replied:
@isidentical @lysnikolaou thoughts?

isidentical replied:
> One of the reasons is that the : defined the format specifier and we need to parse that as a string but if we allow that chunk to be tokenised normally it will never match sone of the allowed format specifiers because those are invalid combinations of alphanumeric characters.

I guess that's true. I was thinking about how we can circumvent it, but without a direct feedback (or at least coupling with) from the parser (which I know we don't want to do) I don't see a way around it. I also agree with @lysnikolaou that this would not be a particularly good example, so I'd be inclined to change this example (or drop it).

pablogsal replied:
I guess we could do something like what we do with the f-string debug expressions and grab the source directly even if is a lot of random tokens and then build a string with that. We would need a special parser rule than can eat any number of tokens until a given one, which is possible but I don’t know if is worth it

isidentical replied:
> I guess we could do something like what we do with the f-string debug expressions and grab the source directly even if is a lot of random tokens and then build a string with that.

That's true too! I didn't think about it purely because I was thinking about unlike debug expressions the format spec is not a formal python expression.

> We would need a special parser rule than can eat any number of tokens until a given one, which is possible but I don’t know if is worth it

In the event that we also formalize the ['format specification mini-language'](https://docs.python.org/3/library/string.html#formatspec) right in the parser for f-strings, it might be worth it (so we wouldn't do weird any-token eater rules) but I'd recommend it as a subsequent (and minor) improvement which bases itself on this. 

My vote is for deferring this case (particularly where a top-level colon is used, [assuming this doesn't affect colon-equal?]) for this PEP to the future with a recommendation/mention that it is possible to do but might require either the any-token approach you mentioned or a formalization of the format spec language in the grammar which would be a separate feature that'd need this PEP.

lysnikolaou replied:
> We would need a special parser rule than can eat any number of tokens until a given one, which is possible but I don’t know if is worth it

I'm probably also for "it's not worth it", so dropping this altogether (at least for now) makes sense to me.

> In the event that we also formalize the ['format specification mini-language'](https://docs.python.org/3/library/string.html#formatspec) right in the parser for f-strings, it might be worth it (so we wouldn't do weird any-token eater rules) but I'd recommend it as a subsequent (and minor) improvement which bases itself on this.

I would be +1 on this exactly as presented. We can start off with what we have now and improve over it later.

pablogsal replied:
Ok, let's move it then to "rejected ideas" or something similar. Are you ok with that?

isidentical replied:
Agreed.

lysnikolaou replied:
Yup, let's move to rejected ideas.

pablogsal replied:
Done

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,247 @@
+can be lifted by giving f-literals a regular grammar without exceptions
+and implementing it using dedicated parse code.
+
+The other problems that f-strings have is that the current implementation in
+CPython relies on the existing string parsing machinery and a post processing of
+its tokens. This has the following problems:
```
I guess the point here is that f-string get tokenised as `STRING` tokens, but I had to go through the sentence twice to understand it. Maybe something like this would be better:

```suggestion
The other problems that f-strings have is that the current implementation in
CPython relies on tokenising f-string as `STRING` tokens and a post processing of
these tokens. This has the following problems:
```

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,247 @@
+   to deal with a lot of manual memory management over the original lexer buffers.
+
+#. The f-string parsing code is not able to use the new improved error message mechanisms
+   that the new PEG parser originally introduced in :pep:`617` has allowed. The
```
```suggestion
   that the new PEG parser, originally introduced in :pep:`617`, has allowed. The
```

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,247 @@
+
+#. The f-string parsing code is not able to use the new improved error message mechanisms
+   that the new PEG parser originally introduced in :pep:`617` has allowed. The
+   improvements that these error messages has been greatly celebrated but
```
```suggestion
   improvements that these error messages brought has been greatly celebrated but
```

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,247 @@
+
+* ``FSTRING_START``: This token includes f-string character (``f``/``F``) and the open quote(s).
+* ``FSTRING_MIDDLE``: This token includes everything between two expression braces (``}`` and ``{``).
+* ``FSTRING_END``: This token includes everything after the last expression brace the until the closing quote.
```
```suggestion
* ``FSTRING_END``: This token includes everything after the last expression brace (or the whole literal part
if no expression exists) until the closing quote.
```

lysnikolaou reviewed (commented):
A few initial comments. Will do another round later today and start filling the missing parts.

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,247 @@
+
+TO BE DONE:
+
+* Describe why we are allowing crazy parenthesis nesting.
```
@pablogsal Do you mean crazy f-string nesting here?

pablogsal replied:
Yeah, apologies. I meant why we are allowing crazy "nest f-strings with the same quotes" as in f"{f"hehe"}"

lysnikolaou commented:
> Btw, we may want to change the grammar to the one they seem to use in LibCST if we find it more readable/better:
> 
> https://github.com/Instagram/LibCST/blob/main/native/libcst/src/parser/grammar.rs#L1401
> 
> @isidentical @lysnikolaou What do you think?

I feel their grammar is more readable to be honest, so I'd be in favour of changing ours if you all agree. I'm also okay with having different grammars in the pep vs in the implementation, if they're syntactically equivalent and since the purpose of each is obviously complete different.

isidentical commented:
@pablogsal @lysnikolaou Have you tried out mass testing (as done with pegen during the initial 3.9 migration) of PyPI packages in the sprints? If not, I can probably set it up today to ensure that everything that is parsable with stock 3.11 is also parsable with our fork on a couple of thousand popular packages.

lysnikolaou commented:
> @pablogsal @lysnikolaou Have you tried out mass testing (as done with pegen during the initial 3.9 migration) of PyPI packages in the sprints? If not, I can probably set it up today to ensure that everything that is parsable with stock 3.11 is also parsable with our fork on a couple of thousand popular packages.

Nope, we didn't during the sprints. And I think @pablogsal hasn't done it either. This would be great if you can take it up!

pablogsal commented:
> > @pablogsal @lysnikolaou Have you tried out mass testing (as done with pegen during the initial 3.9 migration) of PyPI packages in the sprints? If not, I can probably set it up today to ensure that everything that is parsable with stock 3.11 is also parsable with our fork on a couple of thousand popular packages.
> 
> 
> 
> Nope, we didn't during the sprints. And I think @pablogsal hasn't done it either. This would be great if you can take it up!

I would recommend to fix the grammar first. If we are you to change to the one used by libcst let's do it before mass testing

isidentical commented:
> Btw, we may want to change the grammar to the one they seem to use in LibCST if we find it more readable/better:

I tried to view them side by side, but still not sure in which way its different than ours? It seems like LibCST's grammar is more granularly split but the essence of the rules (if we inline `_f_spec()` and `_f_conversion()` and move `gstring_middle` directly to the usage) is practically the same with our implementation. Although I may be missing a point, so would love to hear how we can improve our version (or do you mean like splitting stuff and better naming rules, which shouldn't affect the parsing process)?

pablogsal commented:
> Although I may be missing a point, so would love to hear how we can improve our version (or do you mean like splitting stuff and better naming rules, which shouldn't affect the parsing process)?

Good operation! I apologize because I think I misread their grammar actually. Seems that they have `_f_string` which is an alias for the `FSTRING_MIDDLE` token, which is not very interesting. I think they were looping back a bit more elegantly but upon inspection both grammars are equivalent (as equivalent as they can be in PEG).

pablogsal committed:
Apply suggestions from code review

Co-authored-by: Batuhan Taskaya <isidentical@gmail.com>
Co-authored-by: Lysandros Nikolaou <lisandrosnik@gmail.com>

pablogsal commented on pep-3157.txt:
```diff
@@ -0,0 +1,248 @@
+received.
+
+Rationale
+=========
```
This is the biggest section that is left. Please, @isidentical @lysnikolaou indicate what points you want to work on (or push directly to the branch). Otherwise, if non of you have time this week I will redact something by the weekend

isidentical replied:
I've pushed my first draft but am unsure if it completely fits as a "rationale" (let me know if there are particular ideas that we want to emphasize more). The main thesis I've used was the initial PEP's lack of formal definition allowed it to be affected by implementation details of CPython's parser, which we can now lift and reduce the burden on people (by lifting the mental burden of all these rules), tooling authors and CPython maintaince. Some of these stuff are already mentioned in the motivation, so I tried to keep it short but I'd be happy to also add other stuff (if you have any suggestions)

pablogsal replied:
Thanks a lot @isidentical! I will review the text soon. Just for clarification: the purpose of this section is to explain the decisions we have made and why we have made them more than why we are making the pep.

lysnikolaou replied:
Pushed a section with a "rationale" (it's basically cause it's more intuitive) for including backslashes in the expression part.

isidentical committed:
initial rationale

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,282 @@
+        Batuhan Taskaya <batuhan@python.org>,
+        Lysandros Nikolaou <lisandrosnik@gmail.com>
+Sponsor: Pablo Galindo <pablogsal@python.org>
+Discussions-To: python-dev@python.org
```
Nit: maybe have discussions on Discourse instead? What have you been doing with your latest PEPs?

pablogsal replied:
Ah apologies, yeah discussions are going 100% to discourse 👍

lysnikolaou committed:
Add rationale point about backslashes in expressions

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,308 @@
+
+As an example, ``f'some words {a+b} more words {c+d} final words'`` will be tokenized as::
+
+    FSTRING_START - "f'"
+    FSTRING_MIDDLE - 'some words '
+    LBRACE - '{'
+    NAME - 'a'
+    PLUS - '+'
+    NAME - 'b'
+    RBRACE - '}'
+    FSTRING_MIDDLE
+    - , string='{'
+    NAME - 'c'
+    PLUS - '+'
+    NAME - 'd'
+    RBRACE - '}'
+    FSTRING_END - ' final words'
```
Is there a typo here?

```suggestion
    FSTRING_START - "f'"
    FSTRING_MIDDLE - 'some words '
    LBRACE - '{'
    NAME - 'a'
    PLUS - '+'
    NAME - 'b'
    RBRACE - '}'
    FSTRING_MIDDLE - ' more words '
    LBRACE - '{'
    NAME - 'c'
    PLUS - '+'
    NAME - 'd'
    RBRACE - '}'
    FSTRING_END - ' final words' (without the end quote)
```

Also, regarding the end quotes not being part of the FSTRING_END token, we probably need to explain this, but also, is it really needed? I can't remember why that is the way it is.

pablogsal replied:
I don't think there is a reason, this is something we carried out from the draft. We probably should change it. I think it just makes easier to produce the final string because we don't need to strip the quotes.

pablogsal replied:
> Is there a typo here?

Yep, apologies there was a bad copy paste. Fixing it

pablogsal committed:
Update pep-3157.txt

Co-authored-by: Lysandros Nikolaou <lisandrosnik@gmail.com>

pablogsal committed:
Add more good stuff

pablogsal commented:
@lysnikolaou @isidentical I have completed the pending sections, fixed the pending points in the discussion, fixed the rendering errors and added a bunch of extra clarifications in several sections.

pablogsal commented:
I have pushed a bunch of typo fixes.

pablogsal commented:
If you are ok with this, I will submit the draft to the PEP repo and future fixes will be done there. Please, confirm with :+1:

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,368 @@
+
+This document proposes to lift some of the restrictions originally formulated in
+:pep:`498` and to provide a formalized grammar for f-strings that can be
+integrated into parser directly. The proposed syntactic formalization of
```
```suggestion
integrated into the parser directly. The proposed syntactic formalization of
```

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,368 @@
+provided without providing a formal grammar for f-strings. Additionally, the
+specification contains several restrictions that are imposed so the parsing of
+f-strings could be implemented into CPython without modifying the existing
+lexer. These limitations have been recognized previously and previous attempt
```
```suggestion
lexer. These limitations have been recognized previously and previous attempts
```

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,368 @@
+and implementing it using dedicated parse code.
+
+The other problems that f-strings have is that the current implementation in
+CPython relies on tokenising f-string as `STRING` tokens and a post processing of
```
```suggestion
CPython relies on tokenising f-strings as `STRING` tokens and a post processing of
```

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,368 @@
+   placed in f-strings but provides an intuitive way to manimulate string literals in a
+   more flexible way without to having to fight the limitations of the implementation.
+
+#. Another issue that has feels unintuitive to most is the lack of support for backslashes
```
```suggestion
#. Another issue that has felt unintuitive to most is the lack of support for backslashes
```

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,368 @@
+   (particularly as lexer/parser pipelines need to allow "untokenizing" to
+   support the 'fstrng debugging expressions' and this is specially taxing when
+   arbitrary nesting is allowed). Implementations are therefore free to impose a
+   limit on the nesting depth if they need to. Note that this is not a uncommon
```
```suggestion
   limit on the nesting depth if they need to. Note that this is not an uncommon
```

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,368 @@
+   support the 'fstrng debugging expressions' and this is specially taxing when
+   arbitrary nesting is allowed). Implementations are therefore free to impose a
+   limit on the nesting depth if they need to. Note that this is not a uncommon
+   situation, as the CPython implementation alread imposes several limits all
```
```suggestion
   situation, as the CPython implementation already imposes several limits all
```

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,368 @@
+   situation, as the CPython implementation alread imposes several limits all
+   over the place, including a limit on the nesting depth of parentheses and
+   brackets, a limit on the nesting of the blocks, a limit in the number of
+   branches if 'if' statements, a limit on the number of expressions in
```
```suggestion
   branches in 'if' statements, a limit on the number of expressions in
```

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,368 @@
+=================
+
+As the concept of f-strings is already ubiquitous in the Python community, there is
+no fundamental need for users to learn anything new. However, as the formaized grammar
```
```suggestion
no fundamental need for users to learn anything new. However, as the formalized grammar
```

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,368 @@
+As the concept of f-strings is already ubiquitous in the Python community, there is
+no fundamental need for users to learn anything new. However, as the formaized grammar
+allows some new possibilities, it is important that the formal grammar is added to the
+documentation and explained in detail, explicitly mentioned what constructs are possible
```
```suggestion
documentation and explained in detail, explicitly mentioning what constructs are possible
```

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,368 @@
+no fundamental need for users to learn anything new. However, as the formaized grammar
+allows some new possibilities, it is important that the formal grammar is added to the
+documentation and explained in detail, explicitly mentioned what constructs are possible
+since this PEP is merging to avoid confusion.
```
```suggestion
since this PEP is aiming to avoid confusion.
```

lysnikolaou commented on pep-3157.txt:
```diff
@@ -0,0 +1,368 @@
+be placed inside an f-string expression. In this case the authors think that this work will
+make even more simple to explain this aspect of the language since it can be summarized as:
+
+    You can place any valid Python expression inside an f-string expression".
```
```suggestion
    "You can place any valid Python expression inside an f-string expression".
```

pablogsal commented:
Ah, we need to document that `(f'{3! s}')` is allowed now!

lysnikolaou reviewed (commented):
Last round of typos on my side as well.

lysnikolaou committed:
Last round of typos

lysnikolaou commented:
Pushed a last round of typo fixes. Let's do it!

pablogsal commented:
@isidentical As soon as you give it a last review, I will create a PR against the PEPs repo

isidentical commented on pep-3157.txt:
```diff
@@ -0,0 +1,368 @@
+Author: Pablo Galindo <pablogsal@python.org>,
+        Batuhan Taskaya <batuhan@python.org>,
+        Lysandros Nikolaou <lisandrosnik@gmail.com>
+Sponsor: Pablo Galindo <pablogsal@python.org>
```
Is `Sponsor` applicable when one of the authors is a core dev?

```suggestion
```

isidentical commented on pep-3157.txt:
```diff
@@ -0,0 +1,368 @@
+Type: Standards Track
+Status: Draft
+Content-Type: text/x-rst
+Python-Version: 3.9
```
```suggestion
Python-Version: 3.12
```

isidentical reviewed (approved):
Let's ship it 🚢 , only two minor header hints (and we probably need to change the PEP number to 9999, no?)

lysnikolaou committed:
Apply suggestions from code review

Co-authored-by: Batuhan Taskaya <isidentical@gmail.com>

lysnikolaou committed:
Rename file + fix pep number and creation date
